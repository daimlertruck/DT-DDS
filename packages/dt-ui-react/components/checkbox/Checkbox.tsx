import { ChangeEvent } from 'react';

import { BaseProps } from '../../types';
import { Box } from '../box';

import { CheckboxInputStyled, CheckboxLabelStyled } from './CheckBox.styled';

export interface CheckBoxProps extends BaseProps {
  onChange: (evt: ChangeEvent<HTMLInputElement>) => void;
  isChecked?: boolean;
  isDisabled?: boolean;
}

const CheckBox = ({
  onChange,
  dataTestId,
  children,
  isChecked = false,
  isDisabled = false,
}: CheckBoxProps) => {
  const checkboxId = dataTestId ?? 'checkbox-id';

  return (
    <Box style={{ flexDirection: 'row' }}>
      <CheckboxInputStyled
        id={checkboxId}
        type='checkbox'
        checked={isChecked}
        disabled={isDisabled}
        data-testid={checkboxId}
        onChange={onChange}
      />
      <CheckboxLabelStyled htmlFor={checkboxId}>{children}</CheckboxLabelStyled>
    </Box>
  );
};

export default CheckBox;
