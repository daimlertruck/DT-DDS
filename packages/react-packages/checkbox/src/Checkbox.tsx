import { BaseProps } from '@dt-dds/react-core';
import { Icon } from '@dt-dds/react-icon';
import { useTheme } from '@emotion/react';
import { ChangeEvent, ComponentPropsWithRef, forwardRef } from 'react';

import {
  CheckboxStyled,
  CheckboxInputStyled,
  CheckboxLabelStyled,
  CheckBoxInputWrapper,
} from './Checkbox.styled';

export interface CheckBoxProps
  extends BaseProps,
    ComponentPropsWithRef<'input'> {
  onChange?: (evt: ChangeEvent<HTMLInputElement>) => void;
  isChecked?: boolean;
  isDisabled?: boolean;
}

export const Checkbox = forwardRef<HTMLInputElement, CheckBoxProps>(
  (
    {
      dataTestId = 'checkbox-id',
      onChange,
      children,
      isChecked = false,
      isDisabled = false,
      style,
      ...rest
    }: CheckBoxProps,
    ref
  ) => {
    const theme = useTheme();

    const handleOnChangeTrigger = (
      event: React.ChangeEvent<HTMLInputElement>
    ) => {
      if (isDisabled) {
        return;
      }

      onChange?.(event);
    };

    return (
      <CheckboxStyled
        data-testid={dataTestId}
        isChecked={isChecked}
        isDisabled={isDisabled}
        style={style}
      >
        <CheckBoxInputWrapper>
          <CheckboxInputStyled
            checked={isChecked}
            disabled={isDisabled}
            onChange={handleOnChangeTrigger}
            ref={ref}
            type='checkbox'
            {...rest}
          />
          {isChecked ? (
            <Icon code='check' color={theme.palette.content.contrast} />
          ) : null}
        </CheckBoxInputWrapper>
        {children ? (
          <CheckboxLabelStyled isChecked={isChecked} isDisabled={isDisabled}>
            {children}
          </CheckboxLabelStyled>
        ) : null}
      </CheckboxStyled>
    );
  }
);
