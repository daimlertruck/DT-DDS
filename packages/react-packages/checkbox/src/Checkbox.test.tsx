import { withProviders } from '@dt-ui/react-core';
import { fireEvent, render } from '@testing-library/react';

import { Checkbox } from './Checkbox';

describe('<CheckBox /> component', () => {
  const ProvidedCheckBox = withProviders(Checkbox);

  it('renders checkbox with label', () => {
    const { container } = render(
      <ProvidedCheckBox
        isChecked={false}
        isDisabled={false}
        onChange={() => null}
      >
        Label
      </ProvidedCheckBox>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders checkbox without label', () => {
    const { container } = render(
      <ProvidedCheckBox
        isChecked={false}
        isDisabled={false}
        onChange={() => null}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders checkbox checked', () => {
    const { container } = render(
      <ProvidedCheckBox isChecked isDisabled={false} onChange={() => null} />
    );
    expect(container).toMatchSnapshot();
  });

  it('onChange is triggered', () => {
    const onChangeMock = jest.fn();
    const { getByRole } = render(
      <ProvidedCheckBox
        isChecked={false}
        isDisabled={false}
        onChange={onChangeMock}
      />
    );
    const messageCloseButton = getByRole('checkbox');

    expect(messageCloseButton).toBeInTheDocument();

    fireEvent.click(messageCloseButton);

    expect(onChangeMock).toHaveBeenCalledTimes(1);
  });

  it('renders checkbox disabled', () => {
    const onChangeMock = jest.fn();
    const { container, getByRole } = render(
      <ProvidedCheckBox isChecked={false} isDisabled onChange={onChangeMock} />
    );
    const messageCloseButton = getByRole('checkbox');

    expect(messageCloseButton).toBeInTheDocument();

    fireEvent.click(messageCloseButton);

    expect(onChangeMock).toHaveBeenCalledTimes(0);
    expect(container).toMatchSnapshot();
  });
});
