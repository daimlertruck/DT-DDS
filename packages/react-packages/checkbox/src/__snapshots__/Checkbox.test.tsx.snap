// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBox /> component renders checkbox checked 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  cursor: pointer;
}

.emotion-0>div {
  border-radius: 0px;
  border: none;
  background-color: #00677F;
}

.emotion-2 {
  height: 24px;
  min-height: 24px;
  width: 24px;
  min-width: 24px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
}

.emotion-4:disabled {
  cursor: not-allowed;
}

.emotion-6 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #FFFFFF;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

<div>
  <label
    class="emotion-0 emotion-1"
    data-testid="checkbox-id"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <input
        checked=""
        class="emotion-4 emotion-5"
        type="checkbox"
      />
      <i
        class="emotion-6 emotion-7"
        color="#FFFFFF"
        data-testid="icon"
      >
        check
      </i>
    </div>
  </label>
</div>
`;

exports[`<CheckBox /> component renders checkbox disabled 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  cursor: not-allowed;
}

.emotion-0>div {
  border-radius: 0px;
  border: 1px solid #DEDEDE;
  background-color: #FFFFFF;
}

.emotion-2 {
  height: 24px;
  min-height: 24px;
  width: 24px;
  min-width: 24px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
}

.emotion-4:disabled {
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-0 emotion-1"
    data-testid="checkbox-id"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <input
        class="emotion-4 emotion-5"
        disabled=""
        type="checkbox"
      />
    </div>
  </label>
</div>
`;

exports[`<CheckBox /> component renders checkbox with label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  cursor: pointer;
}

.emotion-0>div {
  border-radius: 0px;
  border: 1px solid #DEDEDE;
  background-color: #FFFFFF;
}

.emotion-0:hover>div {
  background-color: #DEDEDE;
  border: 1px solid #000000;
}

.emotion-2 {
  height: 24px;
  min-height: 24px;
  width: 24px;
  min-width: 24px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
}

.emotion-4:disabled {
  cursor: not-allowed;
}

.emotion-6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  width: 100%;
  padding-left: 8px;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #292929;
}

<div>
  <label
    class="emotion-0 emotion-1"
    data-testid="checkbox-id"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <input
        class="emotion-4 emotion-5"
        type="checkbox"
      />
    </div>
    <span
      class="emotion-6 emotion-7"
    >
      Label
    </span>
  </label>
</div>
`;

exports[`<CheckBox /> component renders checkbox without label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  cursor: pointer;
}

.emotion-0>div {
  border-radius: 0px;
  border: 1px solid #DEDEDE;
  background-color: #FFFFFF;
}

.emotion-0:hover>div {
  background-color: #DEDEDE;
  border: 1px solid #000000;
}

.emotion-2 {
  height: 24px;
  min-height: 24px;
  width: 24px;
  min-width: 24px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
}

.emotion-4:disabled {
  cursor: not-allowed;
}

<div>
  <label
    class="emotion-0 emotion-1"
    data-testid="checkbox-id"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <input
        class="emotion-4 emotion-5"
        type="checkbox"
      />
    </div>
  </label>
</div>
`;
